fluid/neural_machine_translation/transformer/reader.py

```diff
@@ -263,95 +262,41 @@ def load_dict(dict_path, reverse=False):
                     word_dict[line.strip()] = idx
         return word_dict
 
-    def _sample_generator(self):
+    def batch_generator(self):
+        # global sort or global shuffle
+        beg = time.time()
         if self._sort_type == SortType.GLOBAL:
-            if not self._sorted:
-                self._sample_idxs.sort(
-                    key=lambda idx: max(len(self._src_seq_ids[idx]),
-                    len(self._trg_seq_ids[idx] if not self._only_src else 0))
-                )
-                self._sorted = True
+            infos = sorted(
+                self._sample_infos,
+                key=lambda x: max(x[1], x[2]) if not self._only_src else x[1])
```

Comment:
Do we need states to keep the sort results and batches, and thus can avoid extra processing and batch generation